base/system/winlogon/rpcserver.c

```diff
@@ -497,22 +600,56 @@ BaseInitiateSystemShutdownEx(
     BOOLEAN bRebootAfterShutdown,
     ULONG dwReason)
 {
+    HANDLE hThread;
+
     TRACE("BaseInitiateSystemShutdownEx()\n");
     TRACE("  Message: %wZ\n", lpMessage);
     TRACE("  Timeout: %lu\n", dwTimeout);
     TRACE("  Force: %d\n", bForceAppsClosed);
     TRACE("  Reboot: %d\n", bRebootAfterShutdown);
     TRACE("  Reason: %lu\n", dwReason);
 
-//    return ERROR_SUCCESS;
+    //FIXME: Verify that the caller actually has the correct privileges
 
-    /* FIXME */
-    return ExitWindowsEx((bRebootAfterShutdown ? EWX_REBOOT : EWX_SHUTDOWN) |
-                         (bForceAppsClosed ? EWX_FORCE : 0),
-                         dwReason);
+    if (_InterlockedCompareExchange8((volatile char*)&g_bShuttingDown, TRUE, FALSE) == TRUE)
+        return ERROR_SHUTDOWN_IN_PROGRESS;
+    if (lpMessage && lpMessage->Length && lpMessage->Buffer)
+    {
+        g_ShutdownParams.usMessage.Buffer = HeapAlloc(GetProcessHeap(), 0, lpMessage->Length+sizeof(UNICODE_NULL));
+        if (g_ShutdownParams.usMessage.Buffer == NULL)
+        {
+            g_bShuttingDown = FALSE;
+            return GetLastError();
+        }
+        RtlInitEmptyUnicodeString(&(g_ShutdownParams.usMessage), g_ShutdownParams.usMessage.Buffer, lpMessage->Length+sizeof(UNICODE_NULL));
+        RtlCopyUnicodeString(&(g_ShutdownParams.usMessage), (PUNICODE_STRING)lpMessage);
+    }
+    else
+    {
+        RtlInitEmptyUnicodeString(&(g_ShutdownParams.usMessage), NULL, 0);
+        HeapFree(GetProcessHeap(), 0, g_ShutdownParams.usMessage.Buffer);
```

Comment:
This HeapFree() call is buggy.

Reply:
? What's the fix? I don't have much time right now.

Reply:
To remove the HeapFree() but keep the RtlInitEmptyUnicodeString() call.